mm2src/common/file_lock.rs

```diff
@@ -0,0 +1,98 @@
+use crate::now_ms;
+use std::path::Path;
+
+pub struct FileLock<T: AsRef<Path>> {
+    /// Filesystem path of the lock file.
+    lock_path: T,
+    /// The time in seconds after which an outdated lock file can be removed.
+    ttl_sec: u64,
+}
+
+/// Records timestamp to a file contents.
+fn touch(path: &dyn AsRef<Path>, timestamp: u64) -> Result<(), String> {
+    std::fs::write(path.as_ref(), timestamp.to_string()).map_err(|e| ERRL!("{:?}", e))
+}
+
+/// Attempts to read timestamp recorded to a file
+fn read_timestamp(path: &dyn AsRef<Path>) -> Result<u64, String> {
+    match std::fs::read_to_string(path) {
+        Ok(content) => content.parse().map_err(|e| ERRL!("{:?}", e)),
+        Err(e) => ERR!("{:?}", e)
+    }
+}
+
+impl<T: AsRef<Path>> FileLock<T> {
+    pub fn lock(lock_path: T, ttl_sec: u64) -> Result<Option<FileLock<T>>, String> {
+        match std::fs::OpenOptions::new().write(true).create_new(true).open(lock_path.as_ref()) {
+            Ok(_) => {
+                let file_lock = FileLock { lock_path, ttl_sec };
+                try_s!(file_lock.touch());
+                Ok(Some(file_lock))
+            },
+            Err(ref ie) if ie.kind() == std::io::ErrorKind::AlreadyExists => {
+                // See if the existing lock is old enough to be discarded.
+                let lm = match read_timestamp(&lock_path) {
+                    Ok(lm) => lm,
+                    Err(ie) => return ERR!("Error checking {:?}: {}", lock_path.as_ref(), ie)
+                };
+                if now_ms() / 1000 - lm > ttl_sec {
```

Comment:
Let me remind you that `now_ms() / 1000 - lm` can be affected by the time going backwards.
i64, f64 or `lm > now_ms() / 1000 + ttl_sec`

Reply:
Yes, thanks! Worth to fix it in gstuff also as I copy-pasted this part from there (https://github.com/ArtemGr/gstuff.rs/blob/master/gstuff.rs#L551) 🙂

Reply:
gstuff is good, it uses f64 and is not affected. I'm preferring f64 because it allows for these natural-looking comparisons (using seconds and substracting in any direction)

Reply:
True about gstuff, my brain was wishful thinking possibly because of Friday evening 🙂